pkg/controllers/management/secretmigrator/clusters.go

```diff
@@ -1051,104 +442,771 @@ func (m *Migrator) Cleanup(secretName string) error {
+			}
+			if !apierrors.IsNotFound(err) {
+				if credentialSecret, ok := m["credentialSecret"]; ok && credentialSecret != nil {
+					continue
+				}
+				logrus.Tracef("[secretmigrator] migrating secrets for %s pipeline config in cluster %s", model.BitbucketServerType, clusterCopy.Name)
+				bbserver := &apiprjv3.BitbucketServerPipelineConfig{}
+				if err = mapstructure.Decode(m, bbserver); err != nil {
+					logrus.Errorf("[secretmigrator] failed to migrate secrets for %s pipeline config in cluster %s, will retry: %v", model.BitbucketServerType, clusterCopy.Name, err)
+					return clusterResult, err
+				}
+				secret, err := h.migrator.CreateOrUpdateSourceCodeProviderConfigSecret("", bbserver.PrivateKey, clusterCopy, model.BitbucketServerType)
+				if err != nil {
+					logrus.Errorf("[secretmigrator] failed to migrate secrets for %s pipeline config in cluster %s, will retry: %v", model.BitbucketServerType, clusterCopy.Name, err)
+					return clusterResult, err
+				}
+				if secret != nil {
+					logrus.Tracef("[secretmigrator] secret found for %s pipeline config in cluster %s", model.BitbucketServerType, clusterCopy.Name)
+					bbserver.CredentialSecret = secret.Name
+					bbserver.PrivateKey = ""
+					bbserver.ObjectMeta, bbserver.APIVersion, bbserver.Kind, err = setSourceCodeProviderConfigMetadata(m)
+					if err != nil {
+						logrus.Errorf("[secretmigrator] failed to migrate secrets for %s pipeline config in cluster %s, will retry: %v", model.BitbucketServerType, clusterCopy.Name, err)
+						deleteErr := h.migrator.secrets.DeleteNamespaced(SecretNamespace, secret.Name, &metav1.DeleteOptions{})
+						if deleteErr != nil {
+							logrus.Errorf("[secretmigrator] encountered error while handling migration error: %v", deleteErr)
+						}
+						return clusterResult, err
+					}
+					_, err = h.sourceCodeProviderConfigs.ObjectClient().Update(bbserver.Name, bbserver)
+					if err != nil {
+						logrus.Errorf("[secretmigrator] failed to migrate secrets for %s pipeline config in cluster %s, will retry: %v", model.BitbucketServerType, clusterCopy.Name, err)
+						deleteErr := h.migrator.secrets.DeleteNamespaced(SecretNamespace, secret.Name, &metav1.DeleteOptions{})
+						if deleteErr != nil {
+							logrus.Errorf("[secretmigrator] encountered error while handling migration error: %v", deleteErr)
+						}
+						return clusterResult, err
+					}
+				}
+			}
 		}
-	}
 
-	return false
-}
+		logrus.Tracef("[secretmigrator] setting cluster condition [%s] and updating cluster [%s]", apimgmtv3.ClusterConditionSecretsMigrated, clusterCopy.Name)
+		apimgmtv3.ClusterConditionSecretsMigrated.True(clusterCopy)
+		clusterCopy, err = h.clusters.Update(clusterCopy)
+		if err != nil {
+			return clusterResult, err
+		}
+		clusterResult = clusterCopy.DeepCopy()
+		// clusterCopy is returned here since it's value will be passed and fields modified without an update
+		return clusterCopy, err
+	})
+	if err != nil {
+		return clusterCopy, err
```

Comment:
Good catch. I think something along those lines will do. However, we don't want to return an empty struct. It needs to be nil or up to date. If it's not nil then it will be passed to subsequent handlers.

Reply:
Looks like if the object is nil, do just returns the object that was passed in anyway. Due to this I opted into just using the passed in cluster pointer var as the clusterResult var. This will cause the original cluster to be returned in the case that no updates were made. I also change it so even if do returns an err, an update is attempted using the returned obj. This is because Do is written with the intention that the updated will be committed since if an error is returned that means the condition is set to False.